src/lib/formats/rpk.cpp

```diff
@@ -0,0 +1,459 @@
+	, m_type(type)
+	, m_filename(filename)
+	, m_hashes(std::move(hashes))
+	, m_length(length)
+{
+}
+
+
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
+
+util::rpk_socket::~rpk_socket()
+{
+}
+
+
+//-------------------------------------------------
+//  read_file
+//-------------------------------------------------
+
+std::error_condition util::rpk_socket::read_file(std::vector<std::uint8_t> &result) const
+{
+	// find the file
+	if (m_rpk.zipfile().search(m_filename, false) < 0)
+		return rpk_reader::error::INVALID_FILE_REF;
+
+	// prepare a buffer
+	result.resize(m_rpk.zipfile().current_uncompressed_length());
+
+	// read the file
+	std::error_condition const ziperr = m_rpk.zipfile().decompress(&result[0], m_rpk.zipfile().current_uncompressed_length());
+	if (ziperr)
+		return ziperr;
+
+	// perform hash checks, if appropriate
+	if (m_hashes.has_value())
+	{
+		util::hash_collection actual_hashes;
+		actual_hashes.compute(&result[0], result.size(), m_hashes->hash_types().c_str());
+
+		if (actual_hashes != m_hashes)
+			return rpk_reader::error::INVALID_FILE_REF; // Hash check failed
+	}
+
+	// success!
+	return std::error_condition();
+}
+
+
+/***************************************************************************
+	RPK EXCEPTION HANDLING
+***************************************************************************/
+
+//-------------------------------------------------
+//  rpk_category - gets the RPK error category instance
+//-------------------------------------------------
+
+std::error_category const &util::rpk_category() noexcept
+{
+	return f_rpk_category_instance;
+}
+
+//-------------------------------------------------
+//  error_message
+//-------------------------------------------------
+
+std::string rpk_category_impl::message(int condition) const
+{
+	const char *result;
+	switch ((util::rpk_reader::error) condition)
+	{
+	case util::rpk_reader::error::NOT_ZIP_FORMAT:
+		result = "Not a RPK (zip) file";
+		break;
+	case util::rpk_reader::error::XML_ERROR:
+		result = "XML format error";
+		break;
+	case util::rpk_reader::error::ZIP_ERROR:
+		result = "Zip file error";
+		break;
+	case util::rpk_reader::error::ZIP_UNSUPPORTED:
+		result = "Unsupported zip version";
+		break;
+	case util::rpk_reader::error::MISSING_RAM_LENGTH:
+		result = "Missing RAM length";
+		break;
+	case util::rpk_reader::error::INVALID_RAM_SPEC:
+		result = "Invalid RAM specification";
+		break;
+	case util::rpk_reader::error::INVALID_RESOURCE_REF:
+		result = "Invalid resource reference";
+		break;
+	case util::rpk_reader::error::INVALID_LAYOUT:
+		result = "layout.xml not valid";
+		break;
+	case util::rpk_reader::error::MISSING_LAYOUT:
+		result = "Missing layout";
+		break;
+	case util::rpk_reader::error::UNKNOWN_PCB_TYPE:
+		result = "Unknown pcb type";
+		break;
+	case util::rpk_reader::error::UNSUPPORTED_RPK_FEATURE:
+		result = "RPK feature not supported";
+		break;
```

Comment:
If you’re going to do this with a switch rather than an array synchronised with the `enum`, use the literal suffix for `std::string`, return values directly, and avoid the extra local and `strlen` that you’re imposing here.  Also, you need to deal with zero being expected to be a no error condition.